Задание 1. Базовое изучение
Изучить представленный набор данных на основе описания его столбцов и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

In [1]:
import pandas as pd

In [4]:
columns = ['surgery', 'age', 'Hospital Number', 'rectal temperature', 'pulse', 'respiratory rate', 'temperature of extremities',
'peripheral pulse', 'mucous membranes', 'capillary refill time', 'pain', 'peristalsis', 'abdominal distension', 'nasogastric tube',
'nasogastric reflux', 'nasogastric reflux PH', 'rectal examination - feces', 'abdomen', 'packed cell volume', 'total protein',
'abdominocentesis appearance', 'abdomcentesis total protein', 'outcome', 'first number is site of lesion', 
'second number is type', 'third number is subtype', 'fourth number is specific code', 'cp_data']

In [16]:
df = pd.read_csv('https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv', sep = ',',
header = 0, names = columns, na_values = '?')

In [17]:
df.head()

surgery  age  Hospital Number  rectal temperature  pulse  respiratory rate  \
0      1.0    1           534817                39.2   88.0              20.0   
1      2.0    1           530334                38.3   40.0              24.0   
2      1.0    9          5290409                39.1  164.0              84.0   
3      2.0    1           530255                37.3  104.0              35.0   
4      2.0    1           528355                 NaN    NaN               NaN   

   temperature of extremities  peripheral pulse  mucous membranes  \
0                         NaN               NaN               4.0   
1                         1.0               1.0               3.0   
2                         4.0               1.0               6.0   
3                         NaN               NaN               6.0   
4                         2.0               1.0               3.0   

   capillary refill time  ...  packed cell volume  total protein  \
0                    1.0  ...                50.0           85.0   
1                    1.0  ...                33.0            6.7   
2                    2.0  ...                48.0            7.2   
3                    2.0  ...                74.0            7.4   
4                    1.0  ...                 NaN            NaN   

   abdominocentesis appearance  abdomcentesis total protein  outcome  \
0                          2.0                          2.0      3.0   
1                          NaN                          NaN      1.0   
2                          3.0                          5.3      2.0   
3                          NaN                          NaN      2.0   
4                          NaN                          NaN      1.0   

   first number is site of lesion  second number is type  \
0                               2                   2208   
1                               2                      0   
2                               1                   2208   
3                               2                   4300   
4                               2                      0   

   third number is subtype  fourth number is specific code  cp_data  
0                        0                               0        2  
1                        0                               0        1  
2                        0                               0        1  
3                        0                               0        2  
4                        0                               0        2  

[5 rows x 28 columns]

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 28 columns):
surgery                           298 non-null float64
age                               299 non-null int64
Hospital Number                   299 non-null int64
rectal temperature                239 non-null float64
pulse                             275 non-null float64
respiratory rate                  241 non-null float64
temperature of extremities        243 non-null float64
peripheral pulse                  230 non-null float64
mucous membranes                  253 non-null float64
capillary refill time             267 non-null float64
pain                              244 non-null float64
peristalsis                       255 non-null float64
abdominal distension              243 non-null float64
nasogastric tube                  196 non-null float64
nasogastric reflux                194 non-null float64
nasogastric reflux PH             53 non-null float64
rectal examination -

In [20]:
for col in df.columns:
    pct_missing = df[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

surgery - 0.3%
age - 0.0%
Hospital Number - 0.0%
rectal temperature - 20.1%
pulse - 8.0%
respiratory rate - 19.4%
temperature of extremities - 18.7%
peripheral pulse - 23.1%
mucous membranes - 15.4%
capillary refill time - 10.7%
pain - 18.4%
peristalsis - 14.7%
abdominal distension - 18.7%
nasogastric tube - 34.4%
nasogastric reflux - 35.1%
nasogastric reflux PH - 82.3%
rectal examination - feces - 34.1%
abdomen - 39.5%
packed cell volume - 9.7%
total protein - 11.0%
abdominocentesis appearance - 54.8%
abdomcentesis total protein - 65.9%
outcome - 0.3%
first number is site of lesion - 0.0%
second number is type - 0.0%
third number is subtype - 0.0%
fourth number is specific code - 0.0%
cp_data - 0.0%


In [28]:
df_ = df[df.columns[[0, 1, 4, 9, 11, 12, 19, 22]]]
# Выбирала признаки, исходя из объективности, точности влияния и достаточности данных

In [29]:
df_

surgery  age  pulse  capillary refill time  peristalsis  \
0        1.0    1   88.0                    1.0          4.0   
1        2.0    1   40.0                    1.0          3.0   
2        1.0    9  164.0                    2.0          4.0   
3        2.0    1  104.0                    2.0          NaN   
4        2.0    1    NaN                    1.0          3.0   
..       ...  ...    ...                    ...          ...   
294      1.0    1  120.0                    2.0          4.0   
295      2.0    1   72.0                    2.0          3.0   
296      1.0    1   72.0                    1.0          4.0   
297      1.0    1  100.0                    1.0          3.0   
298      1.0    1   40.0                    NaN          NaN   

     abdominal distension  total protein  outcome  
0                     2.0           85.0      3.0  
1                     1.0            6.7      1.0  
2                     4.0            7.2      2.0  
3                     NaN            7.4      2.0  
4                     2.0            NaN      1.0  
..                    ...            ...      ...  
294                   NaN           65.0      3.0  
295                   3.0            NaN      3.0  
296                   3.0            6.8      2.0  
297                   3.0            6.0      1.0  
298                   NaN           62.0      3.0  

[299 rows x 8 columns]

Задание 2. Работа с выбросами
В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

In [244]:
df_describe = df_.describe().reset_index()
df_describe

index     surgery        age       pulse  capillary refill time  \
0  count  298.000000  299.00000  299.000000             267.000000   
1   mean    1.395973    1.64214   71.404682               1.303371   
2    std    0.489881    2.17730   28.013422               0.476623   
3    min    1.000000    1.00000   30.000000               1.000000   
4    25%    1.000000    1.00000   48.000000               1.000000   
5    50%    1.000000    1.00000   60.000000               1.000000   
6    75%    2.000000    1.00000   88.000000               2.000000   
7    max    2.000000    9.00000  184.000000               3.000000   

   peristalsis  abdominal distension  total protein     outcome  
0   255.000000            243.000000     266.000000  298.000000  
1     2.913725              2.259259      24.517293    1.550336  
2     0.976298              1.061472      27.509056    0.737967  
3     1.000000              1.000000       3.300000    1.000000  
4     3.000000              1.000000       6.500000    1.000000  
5     3.000000              2.000000       7.500000    1.000000  
6     4.000000              3.000000      57.000000    2.000000  
7     4.000000              4.000000      89.000000    3.000000

In [241]:
df_c.describe()

surgery         age       pulse  capillary refill time  peristalsis  \
count  210.000000  210.000000  210.000000             210.000000   210.000000   
mean     1.400000    1.533333   71.204762               1.280952     2.909524   
std      0.491069    2.000319   27.072665               0.471300     0.976472   
min      1.000000    1.000000   36.000000               1.000000     1.000000   
25%      1.000000    1.000000   50.000000               1.000000     3.000000   
50%      1.000000    1.000000   60.000000               1.000000     3.000000   
75%      2.000000    1.000000   88.000000               2.000000     4.000000   
max      2.000000    9.000000  184.000000               3.000000     4.000000   

       abdominal distension  total protein     outcome  
count            210.000000     210.000000  210.000000  
mean               2.166667      23.693333    1.485714  
std                1.028695      27.365187    0.700161  
min                1.000000       3.300000    1.000000  
25%                1.000000       6.500000    1.000000  
50%                2.000000       7.400000    1.000000  
75%                3.000000      55.750000    2.000000  
max                4.000000      89.000000    3.000000

In [254]:
l = []
for col in df_.columns:
    mode = df_[col].mode()
    l.append(mode[0])
print(l)
#df_describe[8] = l вопрос: как добавить строку к df_describe?

[1.0, 1, 60.0, 1.0, 3.0, 1.0, 6.5, 1.0, 1.0]


ValueError: Length of values does not match length of index

In [35]:
for col in df_.columns:
    val = df_[col].value_counts()
    print(val)

1.0    180
2.0    118
Name: surgery, dtype: int64
1    275
9     24
Name: age, dtype: int64
48.0     28
60.0     25
40.0     18
88.0     12
52.0     12
72.0     11
100.0    11
44.0     11
120.0    10
42.0     10
96.0      8
64.0      8
84.0      8
80.0      8
56.0      7
54.0      7
50.0      6
66.0      5
68.0      5
104.0     5
70.0      4
76.0      4
92.0      4
78.0      3
90.0      3
108.0     3
86.0      3
36.0      3
112.0     3
130.0     2
140.0     2
124.0     2
114.0     2
45.0      2
150.0     2
30.0      2
146.0     1
136.0     1
184.0     1
164.0     1
129.0     1
65.0      1
132.0     1
49.0      1
98.0      1
38.0      1
160.0     1
128.0     1
110.0     1
46.0      1
75.0      1
82.0      1
Name: pulse, dtype: int64
1.0    188
2.0     77
3.0      2
Name: capillary refill time, dtype: int64
3.0    128
4.0     72
1.0     39
2.0     16
Name: peristalsis, dtype: int64
1.0    76
3.0    65
2.0    65
4.0    37
Name: abdominal distension, dtype: int64
6.5     15
7.0     15
7.5 

In [39]:
df_loc = df_.loc[df['age'] == 9]
df_loc

surgery  age  pulse  capillary refill time  peristalsis  \
2        1.0    9  164.0                    2.0          4.0   
8        2.0    9   90.0                    1.0          3.0   
12       2.0    9   92.0                    1.0          3.0   
15       1.0    9  128.0                    2.0          4.0   
22       1.0    9  130.0                    1.0          4.0   
38       1.0    9  146.0                    NaN          NaN   
40       2.0    9  150.0                    NaN          NaN   
54       1.0    9  160.0                    1.0          3.0   
73       1.0    9    NaN                    NaN          NaN   
74       1.0    9  100.0                    2.0          3.0   
81       1.0    9  100.0                    1.0          4.0   
102      1.0    9  140.0                    1.0          3.0   
108      1.0    9   84.0                    1.0          3.0   
134      1.0    9  136.0                    1.0          1.0   
190      2.0    9  100.0                    1.0          1.0   
211      1.0    9  120.0                    2.0          3.0   
226      1.0    9    NaN                    NaN          NaN   
228      1.0    9  120.0                    NaN          1.0   
243      1.0    9  124.0                    1.0          3.0   
251      1.0    9  124.0                    1.0          3.0   
254      1.0    9  184.0                    1.0          1.0   
261      2.0    9   88.0                    NaN          NaN   
274      1.0    9  150.0                    2.0          3.0   
286      2.0    9   84.0                    1.0          NaN   

     abdominal distension  total protein  outcome  
2                     4.0            7.2      2.0  
8                     1.0            6.2      1.0  
12                    2.0            6.1      2.0  
15                    3.0            7.8      2.0  
22                    NaN           70.0      1.0  
38                    NaN            NaN      2.0  
40                    NaN            8.5      1.0  
54                    4.0            NaN      2.0  
73                    NaN            4.9      2.0  
74                    NaN           57.0      3.0  
81                    1.0            5.7      1.0  
102                   2.0            5.3      1.0  
108                   3.0            6.6      2.0  
134                   3.0            4.9      2.0  
190                   1.0            4.7      1.0  
211                   3.0            6.3      1.0  
226                   NaN            6.5      1.0  
228                   NaN           54.0      1.0  
243                   4.0            8.0      1.0  
251                   4.0            7.6      2.0  
254                   3.0            3.3      2.0  
261                   NaN            5.9      1.0  
274                   2.0            6.2      2.0  
286                   NaN            5.0      1.0

In [73]:
df_loc.describe()

surgery   age       pulse  capillary refill time  peristalsis  \
count  24.000000  24.0   22.000000              18.000000    18.000000   
mean    1.250000   9.0  123.363636               1.277778     2.777778   
std     0.442326   0.0   28.628234               0.460889     1.060275   
min     1.000000   9.0   84.000000               1.000000     1.000000   
25%     1.000000   9.0  100.000000               1.000000     3.000000   
50%     1.000000   9.0  124.000000               1.000000     3.000000   
75%     1.250000   9.0  144.500000               1.750000     3.000000   
max     2.000000   9.0  184.000000               2.000000     4.000000   

       abdominal distension  total protein    outcome  
count             15.000000      22.000000  24.000000  
mean               2.666667      13.531818   1.541667  
std                1.112697      19.253012   0.588230  
min                1.000000       3.300000   1.000000  
25%                2.000000       5.400000   1.000000  
50%                3.000000       6.250000   1.500000  
75%                3.500000       7.750000   2.000000  
max                4.000000      70.000000   3.000000

In [240]:
df_c.describe()

surgery         age       pulse  capillary refill time  peristalsis  \
count  210.000000  210.000000  210.000000             210.000000   210.000000   
mean     1.400000    1.533333   71.204762               1.280952     2.909524   
std      0.491069    2.000319   27.072665               0.471300     0.976472   
min      1.000000    1.000000   36.000000               1.000000     1.000000   
25%      1.000000    1.000000   50.000000               1.000000     3.000000   
50%      1.000000    1.000000   60.000000               1.000000     3.000000   
75%      2.000000    1.000000   88.000000               2.000000     4.000000   
max      2.000000    9.000000  184.000000               3.000000     4.000000   

       abdominal distension  total protein     outcome  
count            210.000000     210.000000  210.000000  
mean               2.166667      23.693333    1.485714  
std                1.028695      27.365187    0.700161  
min                1.000000       3.300000    1.000000  
25%                1.000000       6.500000    1.000000  
50%                2.000000       7.400000    1.000000  
75%                3.000000      55.750000    2.000000  
max                4.000000      89.000000    3.000000

In [86]:
df_Q = df_[(df_.pulse < 100)|(df_.pulse > 144.5)|(df_['total protein'] < 5.4)|(df_['total protein'] > 7.75)]

In [85]:
df_Q

surgery  age  pulse  capillary refill time  peristalsis  \
0        1.0    1   88.0                    1.0          4.0   
1        2.0    1   40.0                    1.0          3.0   
2        1.0    9  164.0                    2.0          4.0   
5        1.0    1   48.0                    1.0          3.0   
6        1.0    1   60.0                    1.0          4.0   
..       ...  ...    ...                    ...          ...   
293      2.0    1   40.0                    1.0          1.0   
294      1.0    1  120.0                    2.0          4.0   
295      2.0    1   72.0                    2.0          3.0   
296      1.0    1   72.0                    1.0          4.0   
298      1.0    1   40.0                    NaN          NaN   

     abdominal distension  total protein  outcome  
0                     2.0           85.0      3.0  
1                     1.0            6.7      1.0  
2                     4.0            7.2      2.0  
5                     3.0            7.0      1.0  
6                     2.0            8.3      2.0  
..                    ...            ...      ...  
293                   1.0           67.0      1.0  
294                   NaN           65.0      3.0  
295                   3.0            NaN      3.0  
296                   3.0            6.8      2.0  
298                   NaN           62.0      3.0  

[264 rows x 8 columns]

In [97]:
df_pulse = df_[(df_.pulse < 100)|(df_.pulse > 144.5)]
df_pulse

surgery  age  pulse  capillary refill time  peristalsis  \
0        1.0    1   88.0                    1.0          4.0   
1        2.0    1   40.0                    1.0          3.0   
2        1.0    9  164.0                    2.0          4.0   
5        1.0    1   48.0                    1.0          3.0   
6        1.0    1   60.0                    1.0          4.0   
..       ...  ...    ...                    ...          ...   
292      1.0    1   78.0                    1.0          3.0   
293      2.0    1   40.0                    1.0          1.0   
295      2.0    1   72.0                    2.0          3.0   
296      1.0    1   72.0                    1.0          4.0   
298      1.0    1   40.0                    NaN          NaN   

     abdominal distension  total protein  outcome  
0                     2.0           85.0      3.0  
1                     1.0            6.7      1.0  
2                     4.0            7.2      2.0  
5                     3.0            7.0      1.0  
6                     2.0            8.3      2.0  
..                    ...            ...      ...  
292                   NaN           62.0      3.0  
293                   1.0           67.0      1.0  
295                   3.0            NaN      3.0  
296                   3.0            6.8      2.0  
298                   NaN           62.0      3.0  

[230 rows x 8 columns]

In [104]:
df_pulse1 = df_pulse[(df_pulse.age == 1)&(df_pulse.pulse.between(30,40))]
df_pulse1

surgery  age  pulse  capillary refill time  peristalsis  \
1        2.0    1   40.0                    1.0          3.0   
52       2.0    1   40.0                    1.0          NaN   
75       1.0    1   38.0                    1.0          3.0   
80       1.0    1   40.0                    1.0          3.0   
86       2.0    1   40.0                    1.0          NaN   
99       1.0    1   40.0                    1.0          3.0   
101      2.0    1   40.0                    1.0          NaN   
110      1.0    1   40.0                    1.0          3.0   
111      1.0    1   40.0                    1.0          3.0   
121      2.0    1   40.0                    1.0          1.0   
129      1.0    1   30.0                    NaN          NaN   
132      2.0    1   40.0                    1.0          3.0   
165      1.0    1   40.0                    1.0          1.0   
189      1.0    1   40.0                    1.0          3.0   
202      2.0    1   36.0                    1.0          3.0   
214      1.0    1   40.0                    1.0          2.0   
217      1.0    1   40.0                    1.0          2.0   
231      1.0    1   30.0                    NaN          NaN   
241      1.0    1   36.0                    1.0          3.0   
259      2.0    1   40.0                    1.0          3.0   
275      1.0    1   36.0                    1.0          2.0   
277      2.0    1   40.0                    1.0          1.0   
293      2.0    1   40.0                    1.0          1.0   
298      1.0    1   40.0                    NaN          NaN   

     abdominal distension  total protein  outcome  
1                     1.0            6.7      1.0  
52                    NaN            6.4      1.0  
75                    2.0           68.0      1.0  
80                    2.0            NaN      1.0  
86                    NaN            7.0      3.0  
99                    NaN           66.0      1.0  
101                   NaN           57.0      1.0  
110                   2.0           58.0      1.0  
111                   1.0           56.0      1.0  
121                   1.0           62.0      1.0  
129                   NaN            NaN      2.0  
132                   1.0            5.7      1.0  
165                   1.0            5.9      1.0  
189                   1.0            6.7      1.0  
202                   1.0            5.7      1.0  
214                   1.0            7.0      1.0  
217                   NaN            NaN      3.0  
231                   NaN            7.7      1.0  
241                   3.0            5.9      2.0  
259                   2.0           69.0      1.0  
275                   2.0           75.0      3.0  
277                   1.0            6.8      1.0  
293                   1.0           67.0      1.0  
298                   NaN           62.0      3.0

In [107]:
df_pulse2 = df_pulse[(df_pulse.age == 1)&(df_pulse.pulse.between(20,25))]
df_pulse2

Empty DataFrame
Columns: [surgery, age, pulse, capillary refill time, peristalsis, abdominal distension, total protein, outcome]
Index: []

In [111]:
df_protein = df_[(df_['total protein'] < 5.4)|(df_['total protein'] > 7.75)]

In [112]:
df_protein

surgery  age  pulse  capillary refill time  peristalsis  \
0        1.0    1   88.0                    1.0          4.0   
6        1.0    1   60.0                    1.0          4.0   
10       2.0    1   72.0                    1.0          1.0   
13       1.0    1   76.0                    1.0          4.0   
15       1.0    9  128.0                    2.0          4.0   
..       ...  ...    ...                    ...          ...   
289      2.0    1   45.0                    1.0          1.0   
292      1.0    1   78.0                    1.0          3.0   
293      2.0    1   40.0                    1.0          1.0   
294      1.0    1  120.0                    2.0          4.0   
298      1.0    1   40.0                    NaN          NaN   

     abdominal distension  total protein  outcome  
0                     2.0           85.0      3.0  
6                     2.0            8.3      2.0  
10                    2.0            7.8      1.0  
13                    1.0           81.0      1.0  
15                    3.0            7.8      2.0  
..                    ...            ...      ...  
289                   NaN           58.0      1.0  
292                   NaN           62.0      3.0  
293                   1.0           67.0      1.0  
294                   NaN           65.0      3.0  
298                   NaN           62.0      3.0  

[126 rows x 8 columns]

In [117]:
df_protein1 = df_protein[(df_protein['total protein'].between(6, 7.5))]
df_protein1

Empty DataFrame
Columns: [surgery, age, pulse, capillary refill time, peristalsis, abdominal distension, total protein, outcome]
Index: []

In [118]:
df_protein2 = df_protein[(df_protein['total protein'] > 7.75)]
df_protein2

surgery  age  pulse  capillary refill time  peristalsis  \
0        1.0    1   88.0                    1.0          4.0   
6        1.0    1   60.0                    1.0          4.0   
10       2.0    1   72.0                    1.0          1.0   
13       1.0    1   76.0                    1.0          4.0   
15       1.0    9  128.0                    2.0          4.0   
..       ...  ...    ...                    ...          ...   
289      2.0    1   45.0                    1.0          1.0   
292      1.0    1   78.0                    1.0          3.0   
293      2.0    1   40.0                    1.0          1.0   
294      1.0    1  120.0                    2.0          4.0   
298      1.0    1   40.0                    NaN          NaN   

     abdominal distension  total protein  outcome  
0                     2.0           85.0      3.0  
6                     2.0            8.3      2.0  
10                    2.0            7.8      1.0  
13                    1.0           81.0      1.0  
15                    3.0            7.8      2.0  
..                    ...            ...      ...  
289                   NaN           58.0      1.0  
292                   NaN           62.0      3.0  
293                   1.0           67.0      1.0  
294                   NaN           65.0      3.0  
298                   NaN           62.0      3.0  

[115 rows x 8 columns]

In [121]:
df_protein2[(df_protein2['total protein'] > 61)]

surgery  age  pulse  capillary refill time  peristalsis  \
0        1.0    1   88.0                    1.0          4.0   
13       1.0    1   76.0                    1.0          4.0   
22       1.0    9  130.0                    1.0          4.0   
23       1.0    1   60.0                    1.0          4.0   
28       2.0    1   48.0                    1.0          1.0   
36       1.0    1   72.0                    1.0          3.0   
42       1.0    1  120.0                    1.0          4.0   
43       1.0    1  140.0                    2.0          4.0   
62       2.0    1  130.0                    2.0          4.0   
64       1.0    1   60.0                    1.0          3.0   
65       2.0    1   72.0                    2.0          3.0   
75       1.0    1   38.0                    1.0          3.0   
76       2.0    1   52.0                    1.0          1.0   
89       2.0    1   52.0                    NaN          NaN   
94       1.0    1   48.0                    1.0          3.0   
99       1.0    1   40.0                    1.0          3.0   
107      1.0    1   60.0                    2.0          4.0   
113      1.0    1   44.0                    NaN          4.0   
114      2.0    1    NaN                    1.0          3.0   
115      1.0    1   44.0                    1.0          3.0   
117      1.0    1   78.0                    1.0          3.0   
118      2.0    1   56.0                    1.0          3.0   
121      2.0    1   40.0                    1.0          1.0   
131      NaN    1   48.0                    1.0          3.0   
137      2.0    1   56.0                    1.0          1.0   
143      1.0    1   84.0                    1.0          4.0   
144      2.0    1   80.0                    1.0          3.0   
145      1.0    1   48.0                    1.0          4.0   
148      2.0    1   48.0                    NaN          1.0   
149      2.0    1   44.0                    1.0          3.0   
151      1.0    1  100.0                    2.0          4.0   
153      2.0    1   48.0                    1.0          3.0   
156      1.0    1   60.0                    2.0          4.0   
164      1.0    1  120.0                    1.0          3.0   
171      1.0    1  129.0                    1.0          4.0   
175      1.0    1   60.0                    1.0          3.0   
177      1.0    1   60.0                    2.0          4.0   
182      2.0    1   56.0                    1.0          1.0   
183      2.0    1   68.0                    1.0          1.0   
191      1.0    1   48.0                    1.0          3.0   
194      2.0    1   44.0                    NaN          1.0   
197      2.0    1   60.0                    2.0          3.0   
205      1.0    1   96.0                    2.0          4.0   
207      1.0    1   88.0                    2.0          3.0   
220      1.0    1   92.0                    1.0          4.0   
238      1.0    1   70.0                    1.0          3.0   
253      1.0    1   80.0                    1.0          4.0   
255      1.0    1   72.0                    1.0          1.0   
259      2.0    1   40.0                    1.0          3.0   
275      1.0    1   36.0                    1.0          2.0   
279      1.0    1   48.0                    NaN          1.0   
292      1.0    1   78.0                    1.0          3.0   
293      2.0    1   40.0                    1.0          1.0   
294      1.0    1  120.0                    2.0          4.0   
298      1.0    1   40.0                    NaN          NaN   

     abdominal distension  total protein  outcome  
0                     2.0           85.0      3.0  
13                    1.0           81.0      1.0  
22                    NaN           70.0      1.0  
23                    3.0           65.0      1.0  
28                    1.0           76.0      1.0  
36                    1.0           80.0      1.0  
42                    4.0           67.0      3.0  
43  

Искала выбросы по признаку пульс и общий белок (те, что меньше 25% и больше 75%). Оказалось, что из общего числа строк (299), почти все с аномальными показателями белка или пульса одновременно(264).
В разрезе признака пульс - осталось 230строк. Далее убрала взрослых лошадей с нормальным пульсом (30,40) - 24 строки, и осталось 206 строк. При проверке на пульс, спортивных лошадей не найдено. Показатели пульса говорят о плохом кровообращении у большинства лошадей или о большой погрешности приборов. Но, если учесть, что все лошади больные, то, можно учитывать данный показатель как ессть.  
В разрезе признака белок найдено 126 строк. Из них не найдено строк с нормальным содержанием белка. Если изучить описание строк, где белок больше 7,75 и найти там аномалии, то, можно предположить, что после медианного значения 61 - сильно обезвоженные лошади, т.к. от 25-го до 50-го квантиля большой размах, а норма ниже в 8 раз. Если обратить внимание на признак outcome, то никто из лошадей не умер с белком выше 61. Значит, можно оставить для детального изучения выживаемости лошадей, например. 




Задание 3. Работа с пропусками
Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [124]:
df_.isna().sum()

surgery                   1
age                       0
pulse                    24
capillary refill time    32
peristalsis              44
abdominal distension     56
total protein            33
outcome                   1
dtype: int64

In [238]:
df_c = df_.copy().dropna()


In [243]:
df_c.isna().sum()

surgery                  0
age                      0
pulse                    0
capillary refill time    0
peristalsis              0
abdominal distension     0
total protein            0
outcome                  0
dtype: int64

In [ ]:
Я решила удалить столбцы с пропусками, т.к. при методе describe (df_ и df_c) значения критически не изменились. 